flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java

```diff
@@ -114,8 +115,14 @@ protected void serializeOperatorState(OperatorState operatorState, DataOutputStr
                     operatorState.getSubtaskStates();
             dos.writeInt(subtaskStateMap.size());
             for (Map.Entry<Integer, OperatorSubtaskState> entry : subtaskStateMap.entrySet()) {
-                dos.writeInt(entry.getKey());
-                serializeSubtaskState(entry.getValue(), dos);
+                if (entry.getValue().isFinished()) {
+                    // We store a negative index for the finished subtask. In consideration
+                    // of the index 0, the negative index would start from -1.
+                    dos.writeInt(-(entry.getKey() + 1));
```

Comment:
That's a bit too much magic for my taste :(

It kind of worked for `Operator` where it meant the number of enclosed states. Here I find too complex especially with the offsetting logic. Unfortunately, I think we need to adjust the metadata version.

Reply:
@pnowojski @StephanEwen What do you think about adjusting the metadata format to include the finished flag?

Reply:
Is there no other way around this problem? Creating V4 serialiser would be cleaner, but I think I would be also fine with this hack. Unfortunately I don't have much experience with modifying this code.

Regardless I have one concern. Let's assume subtask `1` ends and it's waiting for final checkpoint, checkpoint `42` happens, it's completed, subtask `1` finishes, checkpoints `43`, `44`... are triggered and completed.

1. We need this flag only for that checkpoint `42`. Previous checkpoints would be still running, checkpoints `43`, and the following ones (assuming we didn't loose `notifyCheckpointComplete()` call) wouldn't contain finished state. Right?
2. And the problem is that if we have even single finished subtasks we have to disallow job graph modifications in front of that subtask, right?
3. In that case, is it enough to write this flag per subtask? What if in checkpoint `44` there was no subtask that was waiting for the final checkpoint? Subtask `1` is long finished. Will we write down the metadata for that subtask `1`, and will we write down this negative number/flag?

Reply:
Hi @pnowojski ~ Sorry I might not fully understand the issue, but for one thing, I think we would only mark subtask as `isFinished = true` in the checkpoint if it is indeed FINISHED when the checkpoint is taken. This is because if it has called `operator.finish()` and is waiting for the final checkpoint, in the following checkpoint (namely `42` ~ `44`), we should still be able to snapshot its state and thus in this case we would not have the issue of discarding part of the keyed state~ ?

Then in the checkpoint, we might have
1. If when taking the checkpoint some subtasks of an operator is finished, we would save a finished flag (in some way) for these subtasks.
2. If all the subtasks of an operator is finished, we would directly save a finished flag for the whole operator.

Since the FINISHED state of a task won't be reverted, thus in all the following checkpoints, it would also be marked as FINISHED~?